### PR TITLE
chore(ts): migrate "logger"

### PR DESCRIPTION
**Pull request by Haroenv**
Had to cast the warning, see https://stackoverflow.com/a/48675307 for a similar question.

### CODE REVIEW

**Comment by francoischalifour** [comment]
Food for thoughts: the logger could be given as a middleware and its cache could be mocked during tests.

File: src/lib/utils/logger.ts [hunk 1]
```diff
@@ -1,23 +1,33 @@
 import noop from './noop';
 
-let deprecate = noop;
-let warn = noop;
-
+type Deprecate<TCallback = (...args: any[]) => any> = (
+  fn: TCallback,
+  message: string
+) => TCallback | void;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/lib/utils/logger.ts]: noop returns void

**Comment by Haroenv** [comment]
@francoischalifour how would that be injected in the widget?

File: src/lib/utils/logger.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (samouss) [reply src/lib/utils/logger.ts]: Indeed but it's incorrect. We should never have assign `deprecate` to `noop`. With `noop` the deprecated function is never called in production. By default it should at least always call the provided function.

File: src/lib/utils/logger.ts [hunk 2]
```diff
@@ -1,23 +1,33 @@
...
+let deprecate: Deprecate = noop;
+
+type Warn = (message: string) => void;
+let warn: Warn = noop;
+
+type Warning = {
+  (condition: boolean, message: string): void;
+  cache: object;
```
[the remark this answers is not in the document]
Comment (Haroenv) [reply src/lib/utils/logger.ts]: I didn't know what the type was, so went with the more generic one, let's change it!

```suggestion
  cache: { [message: string]: boolean };
```

File: src/lib/utils/logger.ts [hunk 3]
```diff
@@ -1,15 +1,33 @@
...
+
+type Warning = {
+  (condition: boolean, message: string): void;
+  cache: { [message: string]: boolean };
+};
+
+/**
+ * Logs a warning when this function is called, in development environment only.
+ */
+let deprecate: Deprecate = fn => fn;
```
Comment (samouss) [inline src/lib/utils/logger.ts]: 👍

**Comment by francoischalifour** [comment]
@Haroenv It could be passed during `render` for instance:

```ts
const widgetDefinition = {
  render({ helper, state, logger }) {
    logger.warn();
  }
};
```

**Pull request merged**